src/octoprint/util/comm.py

```diff
@@ -3240,10 +3274,55 @@ def _send_loop(self):
 					# are done processing the last fetched queue entry
 					self._send_queue.task_done()
 
-				# now we just wait for the next clear and then start again
-				self._clear_to_send.wait()
+					if ( self._AdvancedOkSupported ):
+						if ( self._state == self.STATE_PRINTING 
+								and gcode in self._SynchronousCommands):
+									self._SynchronousCommand = self._currentLine - 1
+									self._AdvancedOkSendNextLines = self._SynchronousCommand;
+									return False
+						else:
+							self._SynchronousCommand = -1
+							if ( self._state == self.STATE_PRINTING 
+								and self._AdvancedOkSendNextLines > 0
+								and not self._send_queue.resend_active
+								and (self._currentLine-1) < self._AdvancedOkSendNextLines
+								and self._continue_sending() ):
+								return True
+
+				return False
 			except:
 				self._logger.exception("Caught an exception in the send loop")
+
+		self._clear_to_send.wait()
+
+		while self._send_queue_active:
+
+			if _send_one_from_queue():
+				continue
+
+			if not self._send_queue_active:
+				break
+
+			if ( self._AdvancedOkSupported ):
+				if (self._send_queue.resend_active
+					or self._SynchronousCommand >0
+					or self._AdvancedOkSendNextLines < 0
+					):
+					# now we just wait for the next clear and then start again
+					while True:
```

Comment:
This feels *very* dangerous and is a huge deadlock waiting to happen. It *shouldn't* happen as long as sensible data is being exchanged, but that's not something we can count on with finicky cabling and controllers. This needs *some* kind of fallback to ensure we don't get stuck in this loop forever.

Reply:
you are right, self._advanced_ok_last_line must be retested to -1 at communication timeout, if we loose synchronization